python/pyarrow/_gcsfs.pyx

```diff
@@ -151,17 +159,17 @@ cdef class GcsFileSystem(FileSystem):
     def _reconstruct(cls, kwargs):
         return cls(**kwargs)
 
-    def _expiration_datetime_from_options(self):
+    def _expiration_ns_from_options(self):
         expiration_ns = TimePoint_to_ns(
             self.gcsfs.options().credentials.expiration())
         if expiration_ns == 0:
             return None
-        return datetime.fromtimestamp(expiration_ns / 1e9)
```

Comment:
@pitrou this is where I think we end up losing precision.  When I looked at the delta from  logs in C++ (see prior commits), I might have been doing the math wrong, but I think it was about a 6 hour delta (not just a microseconds vs nanoseconds issue).

Reply:
Can you try to follow how `FileInfo.mtime` is implemented?

Reply:
Note a 6 hour delta is not a precision issue but an accuracy issue.

Reply:
Micah are you GMT+06 and assuming timezone issue because of that?

Reply:
I think I was missing passing the timezone which makes sense, i've reverted prior work around and I'll kick off the problematic crossbow build again.  I think most of our CI instances must run with default timezone as UTC?  I'm personally in UTC-7

Reply:
Perhaps GCS returns in local time? https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/gcsfs.cc#L616

Reply:
this wasn't an mttime issue,  the issue was with this code and that I wasn't passing in UTC to convert back to datetime.  passing in timezone.utc fixed the mac failure.

Reply:
GCS (the service) returns a RFC 3339 timestamp, always in the UTC (Zulu) timezone.  The C++ client library converts that to a `std::chrono::system_clock::time_point`, always using UTC.

I think the issue here is that on macOS `std::chrono::system_block::time_point` does not have nanosecond precision.  It only has microsecond precision:

https://stackoverflow.com/questions/65397041/apple-clang-why-can-i-not-create-a-time-point-from-stdchrononanoseconds

Consider relaxing the test condition to accept timestamps within a range.

Reply:
this issue is solved, i was reconstructing a datetime incorrectly.